### ISSUE TITLE
How is discovery supposed to work?

### ISSUE THREAD

**Issue by anentropic** [docker/classicswarm#1902]
I'm confused about discovery

swarm requires some kind of discovery backend... I'm running consul

I launched some containers in my swarm with docker-compose

I can't find anything except the consul node itself registered as services in consul though... maybe I don't know where to look, but shouldn't the 'services' from my docker-compose.yml be reflected as services in consul? or I got the wrong end of the stick?

I'm not clear if I'm supposed to use something like gliderlabs/registrator, or if that's a bit obsolete now but my swarm is not working properly, or what

ultimately I want to get consul-template working so I can configure my nginx proxy from the discovery details I thought were supposed to be stored in consul

**Comment by kunalkushwaha** [comment]
@anentropic Can you please share details of commands you used for configuring swarm with consul as backend.

In general, use [docker-machine](https://github.com/docker/machine/releases) for easily configuring nodes with docker-swarm.

Typical steps it requires is.
1. Run Consul container on one host. ( Lets say Consul running on ``192.168.99.100:8500``)
2. Create swarm master host with discovery backed as consul.
e.g. 

```
$ docker-machine create \
  >    -d virtualbox \
  >    --swarm \
  >    --swarm-master \
  >    --swarm-discovery=consul://92.168.99.100:8500 \
  >    --engine-opt="cluster-store=consul://92.168.99.100:8500 \
  >    --engine-opt="cluster-advertise=eth1:3376" \
  >    master
```

3. multiple nodes (swarm enabled) you can add by specifying the flags ``--swarm`` and ``--swarm-discovery``
e.g.

  ```
  $ docker-machine create \
  >    -d virtualbox \
  >    --swarm \
  >    --swarm-discovery="consul://$(docker-machine ip swl-consul):8500" \
  >    --engine-opt="cluster-store=consul://92.168.99.100:8500 \
  >    --engine-opt="cluster-advertise=eth1:3376" \  
  >    node1
  ```

You may further look at [Swarm docs](https://docs.docker.com/swarm/install-w-machine/) & [Discovery options](https://docs.docker.com/swarm/discovery/). 

Hope it helps.

**Comment by bk1te** [comment]
@anentropic maybe you should use something like gliderlabs/registrator on eash swarm worker node?
I don't have services registered too with only swarm

**Comment by anentropic** [comment]
My question is really: what is the scope of 'discovery'?

Initially I assumed it was 'service' discovery, and would register the containers you launched etc.

But I think now it is just 'node' discovery, i.e. for internal use of docker swarm.

I just wanted to confirm this before implementing progrium/registrator ...because this feels like something that should be built-in.

Also docker moves so fast it's hard to tell which tutorials and advice on blogs is already obsolete.

**Comment by kunalkushwaha** [comment]
discovery is for internal swarm usage. Stores information required containers exist in swarm also can be used for overlay network.

Docker docs are always up to date but various blogs may not :)

I think, you shall understand [use case of swarm](https://www.docker.com/products/docker-swarm), then should decide weather it is required in your case or not. I you don't require to create docker host cluster,  instead just want to manage few services on single host, you may simply use `registrator`

I think it is very much dependent on your use case. 
Hope it helps.

**Comment by anentropic** [comment]
I understand the use case of swarm

I don't understand why `gliderlabs/registrator` would be useful on a single-host though... in that case all my services would be in a known location and I wouldn't need any service discovery.

The essential problem is I need to point my DNS to a static IP, so I need to set up an HTTP proxy at a known address, that can then forward traffic to the dynamic address of my web app, depending on which node it has been scheduled to by swarm.

So far the recommended approach seems to be _HAProxy + consul-template + registrator_ ...I think we have confirmed this now.

For registrator to work I have to share the docker socket to it... this really feels to me like something that should included, it's an essential missing piece of the puzzle. Anyway, I see it's not on the [roadmap](https://github.com/docker/swarm/blob/master/ROADMAP.md) ...maybe not as part of swarm itself but is there any plan for a 'batteries included' docker discovery service working alongside swarm and compose?

**Comment by kunalkushwaha** [comment]
>I don't understand why gliderlabs/registrator would be useful on a single-host though... in that case all 
>my services would be in a known location and I wouldn't need any service discovery.

AFAIK ``gliderlabs/registrator`` works as service discovery for single host. Containers on single host too can be recreated on error or can be increased/decreased depending on workload. In such case for service discovery ``registrator`` is good fit. 

>The essential problem is I need to point my DNS to a static IP, so I need to set up an HTTP proxy 
>at a known address, that can then forward traffic to the dynamic address of my web app, 
>depending on which node it has been scheduled to by swarm.

You may try to look at Docker [Overlay networking](https://docs.docker.com/engine/userguide/networking/get-started-overlay/).
You can create/publish service and attach/detach containers on the fly. Outside traffic talks to service, So service discovery is automatically done. With v1.10, there is custom IP support.

**Comment by anentropic** [comment]
I thought I understood roughly what was needed but now I'm more confused.

I thought the point of registrator was that it listens to events from the docker daemon and puts information about the containers that are launched into consul... hashicorp consul-template can then be used to watch for that info and generate nginx or haproxy config.  I'd assumed that I would run a proxy server outside of the swarm at a known IP address so I could point the domain name to that, and the proxy would then direct web traffic to the nginx container inside my swarm.

If overlay networking makes all that unnecessary, great. But which IP do I point my domain name to? The swarm master?

**Comment by kunalkushwaha** [comment]
Your understanding on registrator +  consul-template + nginx or haproxy is correct. No confusion :)

for understanding overlay networking with service discovery please go through this document. https://docs.docker.com/engine/userguide/networking/get-started-overlay/

Service discovery is done through ``service name``. Custom IP is for containers not service.

**Comment by anentropic** [comment]
> Custom IP is for containers not service.

what do you want to mean here?

**Comment by kunalkushwaha** [comment]
Hi @anentropic the kind of service discovery (+ load balancing) you are looking for can only be achieved externally  i.e. through Nginx, HAProxy etc at this moment. So your approach is correct.

Sorry for my initial comments regarding service discovery through docker networking. This service discovery is basically for internally services to identify each other over network. So no hardcode IP or names explicitly required to be passed to containers(services) to identify each other. But offer no load-balancing.

Basically docker have an embedded DNS server which enable to containers to discovery each other, if they are on same network [ before that ``--publish-service`` was used. The service name I was referring seems to be obsolete with embedded DNS addition in docker.]

You may go through https://docs.docker.com/engine/userguide/networking/dockernetworks/ guide, current docker networking is exaplined well here.

**Comment by dongluochen** [comment]
The embedded DNS allows you to name/alias containers so containers can be discovered with name. It doesn't help name discovery from external where you need your registration/name service. 

For load balancer integration you can try [interlock](https://docs.docker.com/swarm/swarm_at_scale/04-deploy-app/). There is a nice [blog post](https://bitcontainer.wordpress.com/2015/09/18/scaling-microservices-with-docker-compose-interlock-and-haproxynginx/) explaining the work flow. Interlock can listen on container event streams to add/remove a container from HAProxy configuration.

**Comment by anentropic** [comment]
@dongluochen thanks very much for the tips

**Comment by viveky4d4v** [comment]
I have 3 swarm managers & single consul . The inbuild LB with  DNS works fine when all nodes are up but once a node goes down the DNS entries are not updated & if i try pinging my services the internal LB points to that service as well which went down due to node unhealthy .
FYI - I am using "generic" machine driver .

Is this the time to use "gliderlabs/registrator" image ? 

In a nutshell , DNS entries are not updated if a container/service goes down on a node & LB keeps pointing to the unhealthy service .

**Comment by dongluochen** [comment]
@viveky4d4v I'm a little confused of your setup. Are you referring to Docker Swarm mode in 1.12 release? The built-in LB is in Swarm mode although that setup you don't need consul. You should open issue on https://github.com/docker/docker/issues/ for swarm mode. This repo is for "classic" swarm.

**Comment by viveky4d4v** [comment]
@dongluochen Thanks for the redirection . 

I have a small query - 

Will swarm internal DNS  keep entries for unhealthy containers ? If i have scaled a service to 4 containers & if one of them goes down how the DNS & LB will react to it ?

**Comment by dongluochen** [comment]
@viveky4d4v In Swarm mode, an unhealthy container will be killed by Swarm manager. When the container is removed, it's removed from DNS & LB. 

There is another PR (https://github.com/docker/docker/pull/27279) to make it seamless. Only healthy containers are added to LB. And remove them as soon as they are unhealthy.
